Add eslint rule to find unused properties, data and computed properties

MisRob opened the pull request:
### Summary

1. Implement eslint rule that reports unused properties, data, and computed properties
2. Turn it on and deal with it
3. The last commit contains fix for the following issue: One of the reported errors showed that there was a development mock label in a group exercise lesson report link leading back to a parent lesson page (Coach -> Reports -> Groups –> select a group lesson <lesson_title> -> select an exercise belonging to this lesson and watch backlink label (should be 'Back to "<lesson_title>"' but was 'Back to “Lesson 1"')

**Before**
![before](https://user-images.githubusercontent.com/13509191/55672231-f4fceb00-5898-11e9-9f24-0f20c430374c.gif)

**After**
![after](https://user-images.githubusercontent.com/13509191/55672234-fb8b6280-5898-11e9-9959-05117bf3ca34.gif)

### Reviewer guidance
- Break it :)! Any ideas about possibly untreated situations or edge cases?
- I tried to be careful and check that there is no hidden usage for all properties removed in 
055bfc3 , though I would appreciate if someone had a look
- In 582dcab I removed constants from properties. This could also be solved be leaving them in properties but referencing them using `this`. I guess there's no need to save them to an instance and it's also a bit more visible that they are some common definitions and constants, but I don't have any preference so if someone prefers the second solution, no problem.
- In 9ca2dae I commented out all unused properties related to to commented out template TODO code. I could remove everything but I suppose that it is under development? If it's not the case, I'll remove everything including those template parts.

### References

Partially solves  #4992.

----

As discussed in #4992, as soon as it's ready, I would like to open the rule PR directly in eslint-plugin-vue repo where there's already an issue for this rule opened and proposal accepted.

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.rst

Commit:
Implement no unused properties eslint rule

Commit:
Turn on vue-no-unused-properties

Commit:
Remove unused properties

Commit:
Remove constants from data

Those constants are referenced directly,
they're actually not used as a component
instance data.

Commit:
Fix property casing

Commit:
Comment out unused properties

that are needed for TODO template commented
out code.

Commit:
Fix back link label

rtibbles commented on kolibri/plugins/coach/assets/src/views/reports/ReportsGroupHeader.vue:
```diff
@@ -79,6 +79,7 @@
       group() {
         return this.groupMap[this.$route.params.groupId];
       },
+      /** TODO COACH
```
This looks unfinished?

rtibbles commented on kolibri/plugins/coach/assets/src/views/reports/ReportsGroupReportLessonPage.vue:
```diff
@@ -93,6 +93,7 @@
     components: {},
     mixins: [commonCoach],
     computed: {
+      /** TODO COACH
```
Ditto.

rtibbles commented on packages/eslint-plugin-kolibri/lib/rules/vue-no-unused-properties.js:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  properties.forEach(property => {
+    let kind = 'property';
```
Could maybe make each of these kinds a constant at the top of the file, and then refer to them?
```
const PROPERTY = 'property';
const DATA = 'data';
const COMPUTED_PROPERTY = 'computed property';
```
(this is following our convention of using SCREAMING_SNAKE_CASE constants in the frontend, instead of repeating references to strings).

MisRob replied:
Yes, this will be better, thanks. Done!

rtibbles reviewed (commented):
This is excellent. My read through of the code changes and the linting code make me think this is definitely close to mergeable, but as you suggest, I will also try to break it myself!

indirectlylit commented:
awesome!

> I tried to be careful and check that there is no hidden usage for all properties

One thing to check for is that many of the users of these components will still be passing in the props. For example, we deleted `id` from `TaskProgress` but that prop is [still being passed in](https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/device_management/assets/src/views/SelectContentPage/index.vue#L12).

> I removed constants

We had a [related conversation](https://github.com/learningequality/kolibri/pull/5006#discussion_r256683153) a little while go. The general pattern is that some `thing` in `<script>` needs to be accessed in `<template>`, and then we've tended to do one of:

* put `this.thing = thing;` or `this.$options.thing = thing;` in `mounted()` or somewhere similar
* put `thing` in `data()`
* put `thing() { return thing; }` in computed props

There are lots of these scattered around the code base. I don't think the inconsistency has caused us any particular issues, but might be nice to standardize on a strategy.

> I commented out all unused properties related to to commented out template TODO code

I think some of these relate to some 0.12.y coach tools work @jonboiser is working on. @jonboiser would you mind checking if any of these are actually dead code that should be removed?

Commit:
Define constants for properties groups and their labels

Commit:
Move constants back to data/computed and reference them using this

Commit:
Remove obsolete props from ContentRender parents

- id, contentId, channelId
- related AssessmentWrapper and its parent (ContentPage) cleanup

Commit:
Remove obsolete props from TaskProgress parents

Commit:
Remove obsolete props from DeleteUserModal parents

Commit:
Remove obsolete props from ResetUserPasswordModal parents

MisRob commented:
Thanks for info Devon!

> One thing to check for is that many of the users of these components will still be passing in the props. 

Right, I didn't realize I should also clean this, thanks. It's done in the last four commits.


> We had a related conversation a little while go. The general pattern is that some thing in `<script>` needs to be accessed in `<template>`, and then we've tended to do one of...


OK. It was used only in the script part here but I guess it will be better to keep it consistent and prepared for usage in templates so I moved those constants I removed before back to an instance and referenced them using `this` (2934966).

indirectlylit commented:
> I guess it will be better to keep it consistent and prepared for usage in templates

all right, that's cool. The way you had it before also seemed fine though

MisRob commented:
No problem, I don't have any strong opinion on this particular issue and if this is a general tendency ;)

indirectlylit reviewed (approved):
I'm gonna go ahead and merge this to prevent merge conflicts.

@jonboiser when you get back, could you follow up on the 'TODO' items and flag if there are any that won't be addressed by your work on coach tools?

jtwaleson commented:
This is great!! Are there any plans to add this to eslint-plugin-vue? Could an update be pushed to NPM so we can use these rules via eslint-plugin-kolibri?

MisRob commented:
Hello, we opened a PR in `eslint-plugin-vue` repository one month ago, you can find it here https://github.com/vuejs/eslint-plugin-vue/pull/871 with some more info regarding updates and other, similar, rules.
